DM-34607: First version (#2)
lsst-dm/dmtn-229: DMTN-229.tex

\keywords{Data Management, Rubin Observatory, Legacy Survey of Space and Time, Databases}

\section{Introduction}

The Vera C.\ Rubin Observatory's Legacy Survey of Space and Time (LSST) \cite{2019ApJ...873..111I} will image the entire southern sky every three days and consist of tens of petabytes of raw image data and associated calibration data.
All these files must be tracked, along with the intermediate datasets and output products from pipeline processing, and depending on where the processing occurs the files will be stored on either POSIX file systems or object stores.
The LSST Data Management System (DMS)\cite{2017ASPC..512..279J} is responsible for transferring raw files off the mountain and storing them at the US Data Facility (USDF).
The datasets are then processed by pipelines\cite{2019ASPC..523..521B,2018PASJ...70S...5B} that can be run at different data centers and the results integrated into a unified data release.
This paper will discuss the part of the DMS that abstracts data access from the pipeline algorithms, builds the execution workflow graphs, and allows the processing jobs to be run in large batch processing systems.

\section{The Data Butler}

> The Data Butler (hereafter the ``Butler''), is the system that abstracts the data access details from the pipeline developers.

Reviewer: ```suggestion
The Data Butler (hereafter the ``Butler'') is the system that abstracts the data access details from the pipeline developers.
```

The top requirements for the Butler are:

\begin{itemize}
\item The pipeline writer should not have to know where the data are being read from or written to or what file formats are being used, or even if a file is involved at all.
\item The pipeline writer only has to deal with Python objects.
\item Users should be able to locate the relevant data using common astronomical concepts such as observation identifier, physical filter, patch of sky, or telescope.
\end{itemize}

The current implementation of the Butler is colloquially knowns as the ``Generation 3'' Butler since it is the third implementation during the evolution of the LSST DMS, a project that began nearly 20 years ago.\cite{2006SPIE.6274E..0PK}
The ``Generation 2'' Butler was used for many years but certain limitations in the design and implementation, especially in terms of its relationship with the evolving demands of the pipeline design and need for more flexibility, led to it being completely replaced with a brand new system that shares no code with previous versions.

The two main components of the Butler are the Registry and the Datastore.

> Registry tracks datasets and relationships but has no idea where or how datasets are persisted.

Reviewer: ```suggestion
Registry organizes datasets conceptually and associates them with these astronomical concepts, but it has no idea where or how datasets are persisted.
```

Datastore is responsible for serializing a Python object to a storage system and reading data back from that storage system and recreating the Python object.
There is then a thin layer of code (the responsibility of the \texttt{Butler} Python class itself) providing a unified interface to the user that coordinates the Datastore and Registry interaction to ensure consistency.

\subsection{Defining a Dataset}

A cornerstone of the Butler design is that any dataset can be found by specifying a coordinate in its dimensional space.
These dimensions are generally quantities that can be understood by an astronomer as being relevant for the particular dataset.
For example, raw data can be addressed using dimensions \texttt{instrument}, \texttt{detector}, and \texttt{exposure}.

> A specific dataset must then be specified using a coordinate within that dimensional space, for example: \texttt{instrument="LSSTCam", detector=1, exposure=2024050100023}, will uniquely represent the second detector of the 23rd LSSTCam observation taken on 2024-05-01.

Reviewer: ```suggestion
A specific dataset must then be specified using a coordinate within that dimensional space, for example: \texttt{\{instrument="LSSTCam", detector=1, exposure=2024050100023\}} will uniquely represent the second detector of the 23rd LSSTCam observation taken on 2024-05-01.
```
(the comma after the data coordinate didn't read well to me in terms of taking in the sentence in one go, so I'm trying the braces as a different way to set the data coordinate off from the rest of the text).

In order to determine which dimensions are relevant for a specific dataset, Butler has the concept of a ``dataset type''.

> A dataset type uses a name to refer to the list of relevant dimensions, the ``storage class'' associated with the data itself, and whether or not it is a calibration dataset.

Reviewer: ```suggestion
A dataset type gives a name to the combination of the list of relevant dimensions, an abstraction around the Python type it corresponding, and whether or not it is a calibration dataset.
```
Mostly just minor wordsmithing that you're welcome to ignore or modify further, but I do think it's better to not say "storage class" _just_ yet since it's already a lot to take in at once.

The name of the dataset type itself does not matter, but it is important that the names are used consistently because they are critical when constructing pipelines.
Raw data are commonly given the dataset type \texttt{raw} and the LSST science pipelines use dataset type names such as \texttt{calexp} (calibrated exposure), \texttt{defects} (pixel defect masks), and \texttt{bias} (processed biases), with the intent being for the name to be descriptive enough for a science user to have a good idea as to what type of data it represents.

> The ``storage class'' is how the Butler is informed about the underlying Python type associated with the dataset.

Reviewer: ```suggestion
The abstraction around the Python type is called a ``storage class''.
```

      A \texttt{CHAINED} collection does not itself contain datasets.
\item A \texttt{CALIBRATION} collection is a special type of collection where a calibration dataset (one whose dataset type is marked as a calibration) can be associated with a validity timespan.
\end{itemize}

All these collections are important when building and executing pipelines.

\subsection{The Registry}

Registry can be thought of as a database that can be queried to find out what datasets are available.
We define Registry to have an abstract interface to allow different implementations to co-exist, although we do assume that a SQL-like syntax is supported when constraining queries (effectively a \texttt{WHERE} clause), even if that does not result in SQL being executed in the specific registry implementation.

The primary Registry implementation uses SQLAlchemy \cite{myers2015essential}\footnote{\url{https://www.sqlalchemy.org}} to talk to PostgreSQL\footnote{\url{https://www.postgresql.org}} and SQLite\footnote{\url{https://www.sqlite.org/index.html}} backend databases.

> We do not make use of the Object Relational Mapping interface, and instead use the low-level SQLAlchemy interface to create tables.

Reviewer: ```suggestion
We do not make use of the Object Relational Mapping interface, and instead use the low-level SQLAlchemy "Core" interface to create and interact with tables.
```

> The Registry stores all the dimension information, collection definitions, and knowledge of all the datasets.

Reviewer: ```suggestion
The Registry stores all the dimension information, collection and dataset type definitions, and the associations between these and the actual datasets.
```

All dimension values must be pre-defined before a dataset can be stored with that value, and the dimension records have additional metadata that can be used to constrain dataset queries or to provide more detail as to what a dimension really represents.

> For example, in the default schema the \texttt{exposure} dimension contains metadata including the time of the exposure, the elevation and azimuth, the observation type, and the exposure time, whereas the \texttt{detector} dimension contains metadata including the full name, the raft name, and the detector type.

Reviewer: Just wanted to expand this a bit to include some more registry functionality.  Suggestion text below.

```suggestion
For example, in the default schema the \texttt{exposure} dimension contains metadata including the time of the exposure, the elevation and azimuth, the observation type, and the exposure time, whereas the \texttt{detector} dimension contains metadata including the full name, the raft name, and the detector type.
The tables that hold this dimension metadata exist independently of the datasets they correspond to; for example, while the metadata associated with the \texttt{exposure} dimension metadata is usually derived from the header of a \texttt{raw} dataset, that \texttt{exposure} is also associated the dataset types that represent processing done on that exposure.
The dimension tables -- which often also include foreign key relationships, such as the \texttt{physical\_filter} associated with an \texttt{exposure} -- thus form a data model ``skeleton'' of sorts for the actual datasets, which have no relationships of their own.
In some case those dimension relationship are spatial (some dimensions are associated with regions on the sky) or temporal (associated with time spans).

These dimension relationships allow the \texttt{Registry} to provide a rich query system, based on a custom expression language based on SQL boolean expressions.
These are parsed and translated by \texttt{Registry} into actual SQL queries, and we delegate optimization and execution of these queries almost entirely to the underlying database.
But in contrast to raw SQL the user essentially has to only provide the \texttt{Registry} the equivalent of the \texttt{WHERE} clause, because the \texttt{SELECT} clause is based on what dataset types or dimensions are being requested, and the multi-join \texttt{FROM} clause can be worked internally according to the known dimension relationships, despite those relationships being encoded in YAML configuration rather than code.
The query system also encapsulates the actual database schema from the user, allowing it to be changed for performance reasons without breaking user code.

\subsection{The Datastore}

Datastores are responsible for serializing a Python object to a storage system and reading the data back from the storage system and returning a Python object.

> On put datastore is given a reference to the dataset (encapsulated in the Python class \texttt{DatasetRef}) by Registry (this is the data coordinate, the dataset type, the run collection, and the unique identifier, a UUID, assigned by Registry) and the Python object.

Reviewer: ```suggestion
On \texttt{put} datastore is given a reference to the dataset (encapsulated in the Python class \texttt{DatasetRef}) by Registry (this is the data coordinate, the dataset type, the run collection, and the unique identifier, a UUID, assigned by Registry) and the Python object.
```
I'm not sure we've mentioned by this point that we call writing a dataset a `put`; this formatting helps, but we may want to do more.

\subsubsection{Other Datastores}

The datastore interface does not require datasets to be persisted as files.
For example, we are considering storing pipeline metrics directly into our metrics database \cite{SQR-019,DMTN-203}.
The Butler user would not know whether a metric was being stored as a JSON file in a file datastore or stored directly into a metrics database (or even stored in both places).

\subsection{Client/Server Butler}

The Butler is a Python library that, by default, is set up to use SQLAlchemy to talk to a SQL registry, and will use AWS or Google Cloud credentials to talk to object stores.
This is not a convenient interface if Python is not being used, or if the authorization required to access the database or object store directly is not available to the people trying to access the Butler.
The latter situation is likely to be the situation in the Rubin Science Platform on Google where users will be logged in with Rubin accounts but will not be issued Google credentials.\cite{DMTN-182,2021arXiv211115030O}.

> To solve this problem we are working on a client/server Butler\cite{DMTN-176}.

Reviewer: ```suggestion
To solve this problem we are working on a \texttt{https} client/server Butler\cite{DMTN-176}.
```
For an external audience I think we should clarify this or emphasize that it's a dedicated "Butler server" that's new, as the current butler (especially with PostgreSQL) is really a client of other kinds of servers already.

          register-dcr-subfilters     Add subfilters for chaotic modeling.
          register-instrument         Add an instrument definition to the repository
          register-skymap             Make a SkyMap and add it to a repository.
          remove-collections          Remove one or more non-RUN collections.
          remove-dataset-type         Remove a dataset type definition from a repository.
          remove-runs                 Remove one or more RUN collections.
          retrieve-artifacts          Retrieve file artifacts from a Butler.
          transfer-datasets           Transfer datasets from one butler to another.
          write-curated-calibrations  Add an instrument's curated calibrations.
    \end{verbatim}
    \end{small}
    \caption{The \texttt{butler} command line options and subcommands.}
    \label{fig:butler-cli}
\end{figure}

There are many common operations that should be achievable without needing to write any Python code.
To serve those use cases we provide command-line tooling based on the \texttt{click} Python package.\footnote{\url{https://click.palletsprojects.com}}

> Examples of the current subcommands and options are show in Fig.~\ref{fig:butler-cli}.

Reviewer: ```suggestion
Examples of the current subcommands and options are shown in Fig.~\ref{fig:butler-cli}.
```

It can be very convenient when setting up a new Butler for an instrument to be able to seed it with these curated calibrations without having to locate raw data files and reconstruct them from scratch, and infrastructure is provided to enable that, for example using the \texttt{butler write-curated-calibrations} command line tool.
Generally, when these curated calibrations are stored the Butler converts them from text format to a binary format to make reading them more efficient; this is all handled by the formatter configured for the particular datastore.

\section{The Pipeline System}

\setcounter{footnote}{0}  % Latex only allows 9 footnotes with symbols.

The fundamental unit of work for a pipeline is a \texttt{Task}.
Each \texttt{Task} has a specialized configuration object attached to it (from the \texttt{pex\_config} package\footnote{\url{https://github.com/lsst/pex_config}}) and must provide a \texttt{run()} method that is the method that implements the algorithm.
A \texttt{run()} method alone is not enough to integrate a task into a pipeline.

> To enable that it must be converted into a \texttt{PipelineTask}.

Reviewer: ```suggestion
To enable that it must also satisfy the derived \texttt{PipelineTask} interface.
```

> A \texttt{PipelineTask} is special because it requires that the author declares the required input datasets and output datasets in a way that can be understood by a pipeline graph builder that is attached to a Butler registry.

Reviewer: ```suggestion
A \texttt{PipelineTask} is special because it requires the author to declare the task's ``connections'': the datasets the task will consume as inputs and produce as outputs.
```
We use "connections" later without defining it; this seems like the right place to do it.

> Critically, a \texttt{PipelineTask} must declare the dimensions that it is interested in.

Reviewer: ```suggestion
The connections also include \texttt{PipelineTask} must declare the dimensions over which the task runs.
The task's dimensions do not need to match those of its input or output datasets; for example, a task with dimensions \texttt{\{instrument, exposure\}} that takes an input with dimensions \texttt{\{instrument, exposure, detector\}} is a ``gather'' step that processes all detectors from that exposure together, but processes each exposure independently (and hence possibly in parallel).
```

Newcomers tend to be confused by what a task's dimensions represent, and I hope this example helps.

For example a \texttt{PipelineTask} can declare it works with dimensions \texttt{detector} and \texttt{exposure}, whilst another can work on \texttt{band}, \texttt{tract}, and \texttt{patch} (sky regions).
The input and output datasets are tied in to the Butler by declaring the dataset type name and associated storage class.
The storage class is a proxy for the Python type and thus allows the pipeline author to guarantee that the correct Python type will be given (converting it from another type if required and supported) -- if the storage classes associated with datasets of that dataset type in the target Butler are not compatible with those required by the pipeline the pipeline will not run.

> To support pipeline execution the \texttt{run()} method must now takes parameters that match the connections and must return a \texttt{dict}-like data structure where the keys match the expected outputs.

Reviewer: ```suggestion
To support pipeline execution the \texttt{run()} method must now take parameters that match the connections and must return a \texttt{dict}-like data structure where the keys match the expected outputs.
```

A \texttt{PipelineTask} then has enough information to be able to pull the required datasets from a Butler and store the outputs, although task authors can override that default behavior to allow for more complexity in the Butler interaction.
Connections have options to allow the dataset loading to be deferred (something that is important when co-adding hundreds of images) or to allow multiple datasets to be provided if the dimensionality implies that.

\begin{figure}
  \lstset{language=YAML}
  \begin{lstlisting}
    description: A demo pipeline.
    instrument: lsst.obs.subaru.HyperSuprimeCam
    tasks:
      calibrate:
        class: lsst.pipe.tasks.calibrate.CalibrateTask
        config:
          astrometry.matcher.maxOffsetPix: 300
      characterizeImage: lsst.pipe.tasks.characterizeImage.CharacterizeImageTask
      isr:
        class: lsst.ip.isr.IsrTask
        config:
          doVignette: true
          vignetteValue: 0.0
    \end{lstlisting}
  \caption{An example minimalist HSC pipeline describing single frame processing.}

>   \label{fig:pipeline}

Reviewer: If the style file doesn't prohibit it, we should really make this a fixed-width font (the `butler --help` listing via `verbatim` looks fine).  This is some of the ugliest typesetting I think I've seen LaTeX produce.

Author: I must admit I couldn't really understand how it's possible for a specialist code listing latex package to generate something that was so bad.

Author: Fixed by adding a `\ttfamily` in there.

\end{figure}

Once the \texttt{PipelineTask} code is written they must now be integrated into a pipeline.

> We define pipelines using a YAML text file, an example is shown in Fig.~\ref{fig:pipeline}.

Reviewer: ```suggestion
We define pipelines using a YAML text file; an example is shown in Fig.~\ref{fig:pipeline}.
```

The example pipeline consists of three tasks, with labels \texttt{calibrate}, \texttt{characterizeImage}, and \texttt{isr}, referencing the corresponding \texttt{PipelineTask} Python classes.
The order listed in the YAML file is not important because the classes implementing each task know which dataset types they need and the pipeline builder arranges them in the required order.
The dataset type names for a \texttt{PipelineTask} can be modified in the pipeline definition to allow extra steps to be inserted if desired.

> Each task can have specific configurations given as well as allowing the pipeline executor to take external user overrides.

Reviewer: ```suggestion
Each task can have specific configurations given, and the pipeline executor can also take external user overrides.
```

When an instrument is explicitly specified, the \texttt{Instrument} class can be used to apply configuration overrides to all the tasks where it makes sense for that instrument.
Pipeline definitions can be significantly more complex than in the example, allowing pipelines to be built up using the common recipe and ingredients paradigm\cite{2015A&C.....9...40J,2020ASPC..522..583L} to allow for pipeline reuse.
It is also possible for a user to specify that a subset of a pipeline be executed rather than the entire thing.

> Fig.~\ref{fig:graph}a shows a visualization of the demonstration pipeline shown in Fig.~{fig:pipeline}.

Reviewer: ```suggestion
Fig.~\ref{fig:graph}a shows a visualization of the demonstration pipeline shown in Fig.~\ref{fig:pipeline}.
```

It shows all the input and output datasets (including calibrations) and has determined the order of execution.
What it does not show is any specific datasets attached to each step.
To determine what is to be processed we need to query a Butler and allocate datasets to tasks.
Once a \texttt{PipelineTask} has been assigned a specific set of data, we call that unit of work a ``quantum'', and the set of all quanta to be executed is called a ``Quantum Graph.''

> The user provides an input Butler collection to search for datasets, and a data query to select specific files from that collection.

Reviewer: ```suggestion
The user provides an input Butler collection to search for datasets, in addition to a \texttt{Registry} query expression that affects not just the input datasets but the dimensions out of the tasks and output datasets as well.
Having the \texttt{Registry}'s dimension table skeleton populated before the Quantum Graph is produced allows us to query the database for both input datasets and predicted future output datasets in much the same way.
```

  \label{fig:graph}
  \end{figure}

\section{Integration with Batch Systems}

A Quantum Graph describes the work that the pipeline has to do in order to process all the data, but it does not say how the work should be scheduled or any resources that are required.
To run a Quantum Graph it has to be converted to something that can be executed.
On a single compute node (including a laptop) the \texttt{ctrl\_mpexec} package \footnote{\url{https://github.com/lsst/ctrl_mpexec}} provides a multi-processing executor, commonly run via the \texttt{pipetask} command line, that can run every quantum in the correct order and track all the responses.

When a Quantum Graph consists of thousands of quanta, a single compute node is no longer sufficient.
At that scale we need to use batch computing with tooling such as HTCondor,\cite{10.1002/cpe.938} Pegasus,\cite{10.1016/j.future.2014.10.008} Parsl,\cite{10.1145/3307681.3325400} or PanDA.\cite{10.1088/1742-6596/331/7/072024}

> For these batch systems the Quantum Graph must be translated into the form expected by the target system.

Reviewer 2: The HTCondor, Pegasus, Parsl and PanDA are not batch systems in this context, but workflow systems (e.g., PanDA can run jobs on SLAC's slurm batch system).

A prototype demonstrated the approach on AWS with Pegasus using an S3 object store\cite{2020arXiv201106044B} but that used custom code.
It was soon realized that many of the concepts for translating a Quantum Graph to a specialist graph are common and we developed a new Batch Production System (\texttt{ctrl\_bps}\footnote{\url{https://github.com/ctrl_bps}}) that would provide a framework to allow the aforementioned batch environments to be treated as plugin code.

> The BPS system provides a facility for a Quantum Graph to be converted to a generic workflow graph before being converted to the final form.

Reviewer 2: Not sure of the pattern that has some things capitalized (Quantum Graph) vs lower case (generic workflow graph).

Author: I use "Quantum Graph" as a term because we introduce it as a concept by name. I can switch everything to lower case if you like. Maybe here I could say "converted to generic form supporting workflow system concepts, called the Workflow Graph"

Reviewer 2: That works.  Just wanted to make sure the capitalization was on purpose.

Author: I've made your changes.

> This intermediate graph can take configuration parameters that specify special resource requirements (CPU or memory) required by specific tasks.

Reviewer 2: This intermediate graph includes the special resource requirements.  So maybe "This intermediate graph includes submission configuration parameters"

> It also allows quanta to be clustered for efficiency.

Reviewer 2: Need to clarify the "It".  If "It" is generic workflow graph, then the sentence is incorrect (a clustered quantum graph is converted into a generic workflow graph).  If "It" is the bps submission process, then ok.

Some quanta can be quick, executing in tens of seconds, and not all batch systems are designed to efficiently handle such small payloads.

> Clustering allows for multiple quanta to be combined into a single execution job, this can be particularly efficient if the clustering is designed such that the output of the first quantum is the only input to the next quantum (which is a common scenario for the early stages of many pipelines when a single detector is being characterized).

Reviewer: ```suggestion
Clustering allows for multiple quanta to be combined into a single execution job; this can be particularly efficient if the clustering is designed such that the output of the first quantum is the only input to the next quantum (which is a common scenario for the early stages of many pipelines when a single detector is being characterized).
```

In the initial experiments, each quantum would talk to the main Butler repository when retrieving data at the start of the quantum and then when writing the results out at the end.

> It soon became clear that this approach was not tenable once thousands of jobs run concurrently and they are all trying to ask registry where the specific dataset is located.

Reviewer 2: tense?  "of jobs are running concurrently?   and "to ask registry"  -> "to ask the registry"?

We fixed this problem by changing the way that batch jobs interact with a Butler.

> Initially we created a standalone read-only SQLite Butler registry containing the relevant information for the workflow; that did fix the problem but for very large workflows the SQLite file became larger than the quantum graph and would have to be copied to each node.

Reviewer 2: Wording a little confusing...almost implies that because it got large the SQLite file needed to be copied to each job.
How about something like:
Initially we created a standalone read-only SQLite Butler registry containing the relevant information for the entire workflow.   [I think the merge job needs to be mentioned here to complete the concept]   That did fix the problem with the large numbers of simultaneous connections and updates.  However, as a SQLite file it had to be copied to each job. For very large workflows the SQLite file became large enough that the copying caused its own problems on some systems and at minimum caused time delays.

> A better solution was to use the Quantum Graph itself.

Reviewer 2: Why?  (saves space combining the data rather than repeating some information in both files?  Quantum Graphs are not required to be local and portions of the graph can be read remotely?)  Since I didn't carefully read the document from the beginning, I may have missed it.  I think it is worth repeating here.

> The Quantum Graph knows all the input datasets and all the expected output datasets, it is therefore possible for a Butler to be constructed that uses the Quantum Graph itself as the Registry whilst using the main Datastore.

Reviewer: ```suggestion
The Quantum Graph knows all the input datasets and all the expected output datasets; it is therefore possible for a Butler to be constructed that uses the Quantum Graph itself as the Registry whilst using the main Datastore.
```

To enable this it was necessary to transfer datastore records for existing datasets into the Quantum Graph.
This is necessary since the file locations and formatters are configuration options and therefore the current configuration may not match the configuration when a dataset was stored.

> The output datasets will use the current configuration and are therefore predictable at run time.

Reviewer 2: These last 2 sentences may be too detailed compared to other bits and possibly could be cut.

> Running large workflows in this manner means that Registry is not involved at all during the bulk of the processing.

Reviewer 2: Maybe:  means that no registries (main or job-local)

> Once the main workflow completes, a final job is executed that checks the datastore to see which of the expected files were produced, and then ensures that the associated registry entries are transferred back to the main Registry.

Reviewer 2: Maybe add:  This final job is executed regardless of the workflow success or failure.
(This is one of the more advanced workflow capabilities that we use.)

This significantly improves the scalability of the bulk processing.

> Note though that in this mode we are still writing to the main datastore file system or object store and not using any file management facilities of the workflow system.

Reviewer 2: Minor point.  Not that you may want to go into detail about chained datastores, but it would be nice to not leave the impression that all files have to use the same datastore/location.

> In future we may decide to change the configuration to have the jobs write to a local file system datastore (thereby matching all the formatter and file naming configuration) and have the workflow system move the files to the correct nodes.

Reviewer 2: In future  -> In the future
to the correct nodes ...  The first part of the sentence is about writing (but not reading) and the second part is about moving the inputs, but not the outputs.  So maybe something like:
In the future, we may decide to have jobs read and write to a job-local file system datastore (thereby matching all the formatter and file naming configuration) and have the workflow system move files between the main datastore and the job's datastore

    tasks:
      calibrate:
        class: lsst.pipe.tasks.calibrate.CalibrateTask
        config:
          astrometry.matcher.maxOffsetPix: 300
      characterizeImage: lsst.pipe.tasks.characterizeImage.CharacterizeImageTask
      isr:
        class: lsst.ip.isr.IsrTask
        config:
          doVignette: true
          vignetteValue: 0.0
    \end{lstlisting}
  \caption{An example minimalist HSC pipeline describing single frame processing.}
  \label{fig:pipeline}
\end{figure}

Each \texttt{PipelineTask} provides one data processing step, and may be integrated together into a a processing pipeline.
We define pipelines using a YAML text file; an example is shown in Fig.~\ref{fig:pipeline}.
The example pipeline consists of three tasks, with labels \texttt{calibrate}, \texttt{characterizeImage}, and \texttt{isr}, referencing the corresponding \texttt{PipelineTask} Python classes.
The order listed in the YAML file is not important because the classes implementing each task know which dataset types they need and the pipeline builder arranges them such that the output connections of one task are associated with the corresponding input connections of one or more other tasks.

> Each task can have configurations specified in the pipeline file which override the default values, or overrides may be given through pipeline executor command line interface.

Reviewer 3: After this line, what about:
```suggestion
Each task can have configurations specified in the pipeline file which override the default values, or overrides may be given through pipeline executor command line interface.
This configuration not only supports algorithmic values, but also allows configuring the names used for the datasets in a task's connections.
The ability to configure dataset types provides flexibility to reuse the same \texttt{PipelineTask} with different configurations multiple times in a pipeline, with each configuration outputting a distinct dataset.
Additionally, it also provides flexibility in re-using pipelines.
Different pipelines can be derived from a common base pipeline, but tasks can be added or removed, and the datasets in the task connections can be configured such all the tasks can still be joined into a connected pipeline graph.
```

Author: Thanks for this clarification.

Additionally, some calibrations, such as defect masks, are relatively static, can be represented in a compact text format, and are useful to people outside of a Butler.
We call such datasets ``curated calibrations'' and store them in Git repositories using text file formats and a directory layout that makes the validity ranges clear.
It can be very convenient when setting up a new Butler for an instrument to be able to seed it with these curated calibrations without having to locate raw data files and reconstruct them from scratch, and infrastructure is provided to enable that, for example using the \texttt{butler write-curated-calibrations} command line tool.
Generally, when these curated calibrations are stored the Butler converts them from text format to a binary format to make reading them more efficient; this is all handled by the formatter configured for the particular datastore.

\section{The Pipeline System}

\setcounter{footnote}{0}  % Latex only allows 9 footnotes with symbols.

> The fundamental unit of work for a pipeline is a \texttt{Task}.

Reviewer 3: ```suggestion
Processing data is done with reusable units of code called \texttt{Task}s. 
```

Each \texttt{Task} has a specialized configuration object attached to it (from the \texttt{pex\_config} package\footnote{\url{https://github.com/lsst/pex_config}}) and must provide a \texttt{run()} method that is the method that implements the algorithm.

> A \texttt{run()} method alone is not enough to integrate a task into a pipeline.

Reviewer 3: ```suggestion
\texttt{Task}s can be nested withing each other in a hierarchy to create high level algorithms.
However, the top level \texttt{Task} of any algorithm must satisfy an additional interface, defined as a \texttt{PipelineTask}, which allows it to interact with the \texttt{Butler} for i/o and ordering within a processing pipeline.
```

> To enable that it must also satisfy the derived \texttt{PipelineTask} interface.

Reviewer 3: ```suggestion
```

A \texttt{PipelineTask} is special because it requires the author to declare the task's ``connections'': the datasets the task will consume as inputs and produce as outputs.

> The connections also include \texttt{PipelineTask} must declare the dimensions over which the task runs.

Reviewer 3: ```suggestion
The interface also requires the definition of the task's dimensions, which defines the unit of processing over which the task runs.
```

The task's dimensions do not need to match those of its input or output datasets; for example, a task with dimensions \texttt{\{instrument, exposure\}} that takes an input with dimensions \texttt{\{instrument, exposure, detector\}} is a ``gather'' step that processes all detectors from that exposure together, but processes each exposure independently (and hence possibly in parallel).

> For example a \texttt{PipelineTask} can declare it works with dimensions \texttt{detector} and \texttt{exposure}, whilst another can work on \texttt{band}, \texttt{tract}, and \texttt{patch} (sky regions).

Reviewer 3: ```suggestion
Other examples of \texttt{PipelineTask}  dimentions are \texttt{detector} and \texttt{exposure} which operates on data from an exposure of single detector taken from the telescope , whilst another could be \texttt{band}, \texttt{tract}, and \texttt{patch} (sky regions) which would operate on many intermediate data products produced from previous processing stages that overlap the specified region.
```

The input and output datasets are tied in to the Butler by declaring the dataset type name and associated storage class.
The storage class is a proxy for the Python type and thus allows the pipeline author to guarantee that the correct Python type will be given (converting it from another type if required and supported) -- if the storage classes associated with datasets of that dataset type in the target Butler are not compatible with those required by the pipeline the pipeline will not run.

> To support pipeline execution the \texttt{run()} method must now take parameters that match the connections and must return a \texttt{dict}-like data structure where the keys match the expected outputs.

Reviewer 3: ```suggestion
To support pipeline execution the\texttt{run()} method of a \texttt{PipelineTask} must take parameters that match the input connections and must return a \texttt{dict}-like data structure where the keys match the expected outputs.
```

> A \texttt{PipelineTask} then has enough information to be able to pull the required datasets from a Butler and store the outputs, although task authors can override that default behavior to allow for more complexity in the Butler interaction.

Reviewer 3: ```suggestion
A \texttt{PipelineTask} then has enough information to be able to pull the required datasets from a Butler and store the outputs, although task authors can override that default behavior using the associated config class to allow for more complexity in the Butler interaction.
```

Connections have options to allow the dataset loading to be deferred (something that is important when co-adding hundreds of images) or to allow multiple datasets to be provided if the dimensionality implies that.

\begin{figure}
  \lstset{language=YAML}
  \begin{lstlisting}
    description: A demo pipeline.
    instrument: lsst.obs.subaru.HyperSuprimeCam
    tasks:
      calibrate:
        class: lsst.pipe.tasks.calibrate.CalibrateTask
        config:
          astrometry.matcher.maxOffsetPix: 300
      characterizeImage: lsst.pipe.tasks.characterizeImage.CharacterizeImageTask
      isr:
        class: lsst.ip.isr.IsrTask
        config:
          doVignette: true
          vignetteValue: 0.0
    \end{lstlisting}
  \caption{An example minimalist HSC pipeline describing single frame processing.}
  \label{fig:pipeline}
\end{figure}

> Once the \texttt{PipelineTask} code is written they must now be integrated into a pipeline.

Reviewer 3: ```suggestion
Each \texttt{PipelineTask} provides one data processing step, and may be integrated together into a a processing pipeline.
```

We define pipelines using a YAML text file; an example is shown in Fig.~\ref{fig:pipeline}.
The example pipeline consists of three tasks, with labels \texttt{calibrate}, \texttt{characterizeImage}, and \texttt{isr}, referencing the corresponding \texttt{PipelineTask} Python classes.

> The order listed in the YAML file is not important because the classes implementing each task know which dataset types they need and the pipeline builder arranges them in the required order.

Reviewer 3: ```suggestion
The order listed in the YAML file is not important because the classes implementing each task know which dataset types they need and the pipeline builder arranges them such that the output connections of one task are associated with the corresponding input connections of one or more other tasks.
```

> The dataset type names for a \texttt{PipelineTask} can be modified in the pipeline definition to allow extra steps to be inserted if desired.

Reviewer 3: ```suggestion

```

Author: Is this not true then? I thought that was how we can put things in between two tasks in a pipeline.

Reviewer 3: It's not that it's not true, its just more complicated than this, and I thought it was either out of scope, or would be confusing if it was made too short. If you feel it is a big piece to mention I can try to work something up.

Author: I feel that we should mention that it's possible. An obvious question arising from hard coding dataset types into the task code and then saying that the graph builder works out how to connect tasks up based on the dataset type, is how to add something in between two tasks. We have to say that this is possible so long as dimensions and storage class match expectations and it doesn't require us to edit the python code.

Reviewer 3: In testing we do this (or if say a scientist does some special runs or something) but in general if we intend something to integrate in long term we modify things in the tasks themselves, not only in a yaml.

However it IS configurable there, and possibly more importantly for this functionality, dataset type names can be templateable, and a task can appear multiple times in a pipeline with different configurations (including type names) while keeping the code the same (different types of warps for example)

Reviewer: One major example of this is that we run jointcal and FGCM on HSC but do not run them on ImSim/DC2, and that changes the various inputs to various coaddition tasks.  It looks that's all handled at present by per-instrument config overrides loaded from obs_* packages, but I think it's a case where RFC-775 says we should move them to pipelines post-Gen2.

> Each task can have specific configurations given, and the pipeline executor can also take external user overrides.

Reviewer 3: ```suggestion
Each task can have configurations specified in the pipeline file which override the default values, or overrides may be given through pipeline executor command line interface.
```

> When an instrument is explicitly specified, the \texttt{Instrument} class can be used to apply configuration overrides to all the tasks where it makes sense for that instrument.

Reviewer 3: ```suggestion
When a pipeline specifies an instrument is explicitly, the \texttt{Instrument} class can be used to apply specific configuration overrides to any of the tasks in the pipeline.
```

> Pipeline definitions can be significantly more complex than in the example, allowing pipelines to be built up using the common recipe and ingredients paradigm\cite{2015A&C.....9...40J,2020ASPC..522..583L} to allow for pipeline reuse.

Reviewer 3: ```suggestion
Pipeline definitions can be significantly more complex than in the example. To simplify this process, pipelines may be built up using the common recipe and ingredients paradigm\cite{2015A&C.....9...40J,2020ASPC..522..583L} to allow for pipeline reuse.
```

It is also possible for a user to specify that a subset of a pipeline be executed rather than the entire thing.

Fig.~\ref{fig:graph}a shows a visualization of the demonstration pipeline shown in Fig.~\ref{fig:pipeline}.

> It shows all the input and output datasets (including calibrations) and has determined the order of execution.

Reviewer 3: ```suggestion
It shows all the input and output datasets (including calibrations) and has determined the order of execution, and is referred to as a pipeline graph.
```

What it does not show is any specific datasets attached to each step.

> To determine what is to be processed we need to query a Butler and allocate datasets to tasks.

Reviewer 3: ```suggestion
To determine what is to be processed a graph building algorithm queries a Butler and allocate datasets to tasks.
```

> Once a \texttt{PipelineTask} has been assigned a specific set of data, we call that unit of work a ``quantum'', and the set of all quanta to be executed is called a ``Quantum Graph.''

Reviewer 3: ```suggestion
All of the datasets which correspond to a unique set of task dimensions are bundled together into a unit of work that we call a ``quantum''.  The set of all quanta to be executed and the relations between them and their tasks are called a ``Quantum Graph.''
```

The user provides an input Butler collection to search for datasets, in addition to a \texttt{Registry} query expression that affects not just the input datasets but the dimensions out of the tasks and output datasets as well.

> Having the \texttt{Registry}'s dimension table skeleton populated before the Quantum Graph is produced allows us to query the database for both input datasets and predicted future output datasets in much the same way.

Reviewer 3: ```suggestion
Having the \texttt{Registry}'s dimension table skeleton populated before the Quantum Graph is generated allows us to query the database for both input datasets and predict future output datasets in much the same way.
```

The graph builder then allocates each matching input dataset to a specific Quantum and determines all the expected datasets that will be created during pipeline processing.
During processing the outputs are written to a \texttt{RUN} collection along with log datasets containing all the log messages written by a task, software package versions, and processing metadata (including execution times and CPU usage for \texttt{Task}s).

> In normal usage we then create a new \texttt{CHAINED} collection that combines the input collections and the new output \texttt{RUN} collection to allow a single collection to be queried for all the inputs and outputs of that processing.

Reviewer 3: ```suggestion
In normal usage we then create a new \texttt{CHAINED} collection that combines the input collections and the new output \texttt{RUN} collection to allow a single collection to be queried for all the input and output datasets in that processing.
```

> In this way as processing campaigns proceed chains of collections can be formed representing the final data products.

Reviewer 3: ```suggestion
This procedure produces chains of collections, as processing campaigns proceed, which represent the final data products.
```
